### DISCUSSION TITLE
Any way to get `alembic check` to take pending migrations into account?

### DISCUSSION THREAD

**Discussion by cjw296** [sqlalchemy/alembic#1441] [Usage Questions]
I'm looking for an alembic equivalent of Django's [`makemigrations --check`](https://docs.djangoproject.com/en/5.0/ref/django-admin/#cmdoption-makemigrations-check).

Alembic's `check` command isn't quite this, it appears to be more like Django's [`migrations --check`](https://docs.djangoproject.com/en/5.0/ref/django-admin/#cmdoption-migrate-check).

The use case I have is to ensure that pull requests haven't changed database models in such a way that an alembic migration should have been included in the PR. What's working for folks in this area at the moment?

If I run `alembic check` with correctly generated and up-to-date migrations, I still get:

```
$ alembic check
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
ERROR [alembic.util.messaging] Target database is not up to date.
  FAILED: Target database is not up to date.
```

I'm looking for something that I can run that will take into account the unapplied migrations and only fail if someone forgot to generate one.

**Comment by CaselIT** [comment]
Hi,

there is the check command that should to what you are looking for https://alembic.sqlalchemy.org/en/latest/api/commands.html#alembic.command.check

**Reply by cjw296** [reply]
@CaselIT - not quite, I've added more detail in the original post, but as I said, `alembic check` works like Django's `migrate --check`, I want something that works like `[makemigrations --check](https://docs.djangoproject.com/en/5.0/ref/django-admin/#cmdoption-makemigrations-check)`

**Comment by cjw296** [comment]
On 13/03/2024 10:17, Federico Caselli wrote:
> Hi,
> 
> there is the check command that should to what you are looking for 
> https://alembic.sqlalchemy.org/en/latest/api/commands.html#alembic.command.check <https://alembic.sqlalchemy.org/en/latest/api/commands.html#alembic.command.check>

Almost, but not quite...

https://github.com/sqlalchemy/alembic/discussions/1441

Chris

**Reply by CaselIT** [reply]
If you can't update the target database while performing the check, then no, that can't be done using alembic. And it's not a question of adding support for it, it would need to be another tool entirely

**Reply by cjw296** [reply]
Oh, interesting, is the alembic google group mirrored into these discussions?

Anyway, like I said, Django's `makemigrations --check` manages this, so it's certainly software that can be written - I expect it may well be what folks are _actually_ looking for when they ask for things like `alembic check`...

**Reply by CaselIT** [reply]
> Oh, interesting, is the alembic google group mirrored into these discussions?

No it should be readonly, not sure how you managed to post..

> Anyway, like I said, Django's makemigrations --check manages this, so it's certainly software that can be written - I expect it may well be what folks are actually looking for when they ask for things like alembic check...

I don't doubt that it's possible to do, I'm just saying that doing this would effectively be another library, since it would require to completely rewrite how alembic works and keeps track of changes.

**Reply by zzzeek** [reply]
> 
> Anyway, like I said, Django's `makemigrations --check` manages this, so it's certainly software that can be written - I expect it may well be what folks are _actually_ looking for when they ask for things like `alembic check`...

it is!  if you'd like Alembic to do this, it would be Alembic 2.0 since it would be a complete rewrite of autogenerate and would drastically change everyone's workflow (we'd definitely need to keep the old one around), and you'd implement #117.

The feature of #117 could likely also be done as an add-on to Alembic.

**Reply by zzzeek** [reply]
#117 does not need to change alembic's versioning model and not even the operations model, only how it goes about looking at a MetaData collection and checking if something has changed since a previous iteration of that MetaData collection.

**Reply by cjw296** [reply]
What are people doing in the meantime? What's the timeline for Alembic 2.0?
("autogen from persisted metadata" doesn't sound like what I'm after - I'm basically looking for "does the state of the database plus pending migrations end up with the schema for the model")

**Reply by zzzeek** [reply]
> What are people doing in the meantime? 

they run their DB migrations in their CI environments, which everyone should do anyway

> What's the timeline for Alembic 2.0? 

nonexistent, there are no plans for this

> ("autogen from persisted metadata" doesn't sound like what I'm after - I'm basically looking for "does the state of the database plus pending migrations end up with the schema for the model")

I am sure Django is not looking at the "state of the database" since they dont have any of the reflection capabilities of SQLAlchemy / Alembic.  I would assume they are looking at the current revision number in the DB and assuming that part is present.    

The "autogen from persisted metadata" idea is fairly vague so far and does not include that it would serialize each known state of the DB as diffs, but it could do that just as well.  What it can't do is link individual database migration operations to known schema states, because alembic migrations are open ended, there can be plain `execute()` calls and such.   IIUC, Django's models and migration operations are rigidly constrained to the subset of "django patterns" that are available, so they can link individual migration operations to model changes in a 1-1 fashion.

**Reply by CaselIT** [reply]
> ("autogen from persisted metadata" doesn't sound like what I'm after - I'm basically looking for "does the state of the database plus pending migrations end up with the schema for the model")

if you are you ok with modifying the target database you can achieve this by doing `alembic upgrade heads && alembic check`.

if not then it's currently not possible as explained above

**Reply by cjw296** [reply]
> I am sure Django is not looking at the "state of the database" since they dont have any of the reflection capabilities of SQLAlchemy / Alembic. I would assume they are looking at the current revision number in the DB and assuming that part is present.

A quick skim over [their code](https://github.com/django/django/blob/95ae37839c907d7d030f1387a003a5776593d7d7/django/core/management/commands/makemigrations.py#L105-L261) suggest they load the state from the database and diff the models. I think their migrations table stores more info than is in the `alembic_migrations` table.

> states, because alembic migrations are open ended, there can be plain `execute()` calls and such. IIUC, Django's models and migration operations are rigidly constrained to the subset of "django patterns" that are available,

They also have raw sql stuff in their migrations, so face the same problem, and I think it's reasonably to caveat these as "not covered."

**Reply by CaselIT** [reply]
The issue is not what the migration table contains or not.
Alembic uses schema diffs comparing the database state with the models state. This means that the db needs to have all migration applied to have a meaningful comparison.
Django as far as I know compares a serialized version of the models with the current models state, so no db is required for such operation.

**Reply by zzzeek** [reply]
> They also have raw sql stuff in their migrations, so face the same problem, and I think it's reasonably to caveat these as "not covered."

right but then you have metadata structures unaccounted for, which would then show up as false positives.  if they are building up the state of their migration by summing the individual migration operations and linking them directly to a model, then model structures that were generated by non-standard means will be missed.   in django this is probably fine because the models are extremely limited in what they can include in any case.  SQLAlchemy models can have anything, EXCLUDE constraints, etc.

it would be sufficient to store a structure that represents the actual state of the metadata at different versions, which can produce diffs which then streams into migration operations.    mostly like what we have, just that we dont use reflection to generate the "current DB state".
